### DISCUSSION TITLE
FLOWS: Adding to array before sending to 'Trigger Flow'

### DISCUSSION THREAD

**Discussion by EnzoBitMatrix** [directus/directus#17780] [Archive]
Hi Community,

I have a flow that reads data from a collection, with returns as an array {{$last}}.  
I then send the {{$last}} to a Trigger Flow, which runs a 'Trigger Flow' event for each item in the array - that works perfectly!

Is there a way to send the original {{$trigger.body.keys}} with each item in the array?  For example, my workflow is triggered by a key in a collection, which then looks up another collection (M2M relationship).  The trigger body keys isn't used in the stations collection at all, so it doesn't come up in the array.

![image](https://user-images.githubusercontent.com/22970783/224871721-0d82de29-6b9d-4654-b178-7987ca29d228.png)

For each item in the array, I want to add the original triggering key. The key would be the same for each item in the collection as it doesn't change... Does that make sense?

Thank you in advance for any assistance.

Let's see if I can clean this up :)

I read a collection, and get this result (which is an array... lovely):

![image](https://user-images.githubusercontent.com/22970783/224871299-ca0592ec-4ce0-4ca5-b27b-ce5fd9c3151d.png)

How do I add the triggering key to the body of the array, so it looks like this:

![image](https://user-images.githubusercontent.com/22970783/224871389-0a59ee8c-e2f2-436d-a2f5-81d1319537bb.png)

**Comment by luciandavidescu** [comment]
You'll need to add a  'run script' box and there you can do a for loop (or pretty much everything) with the data.

**Reply by EnzoBitMatrix** [reply]
Ah, thanks.
I assume the code is written in JavaScript?
I am far from a developer, do you have any recommendations on where I might find the resources to learn how to do this?  I have done the traditional goooooogling, but there is so much info that I don't know where to start...

**Reply by luciandavidescu** [reply]
With js indeed. What you need can be done with a for loop. Assuming your array is `arr`...

```
for(let i = 0; i < arr.length; i++) {
  arr[i]."Trigger_key" = "12345";
}
```

Your new `arr` should now include the added key/value pair.

**Reply by EnzoBitMatrix** [reply]
> With js indeed. What you need can be done with a for loop. Assuming your array is `arr`...
> 
> ```
> for(let i = 0; i < arr.length; i++) {
>   arr[i]."Trigger_key" = "12345";
> }
> ```
> 
> Your new `arr` should now include the added key/value pair.

Thank you very much!
I will give it a go.

**Reply by EnzoBitMatrix** [reply]
I really apprc

> > With js indeed. What you need can be done with a for loop. Assuming your array is `arr`...
> > ```
> > for(let i = 0; i < arr.length; i++) {
> >   arr[i]."Trigger_key" = "12345";
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Your new `arr` should now include the added key/value pair.
> 
> Thank you very much! I will give it a go.

I really appreciate the help you've given, but I can't seem to get to work - more than likely becuase I have no idea how to write JS, and trying to follow tuts online isn't cutting it...

This is what I have, and the result...
var arr = "{{$last}}";
module.exports = function(arr) { 
        for(let i = 0; i < arr.length; i++) {
  arr[i]."Trigger_key" = "{{$trigger.body.keys}}";
	};
   return {};
`}

![image](https://user-images.githubusercontent.com/22970783/225476432-ef884f39-9796-44a3-a86b-4a13dc679dc2.png)

Result:
![image](https://user-images.githubusercontent.com/22970783/225476176-86a6c340-6451-4fd6-9a27-1bc8b57ee095.png)

**Reply by luciandavidescu** [reply]
Try this:
```
module.exports = async function(data) {
    var arr = {{$last}}
    var keys = {{$trigger}}.body.keys;
    for(let i = 0; i < arr.length; i++) {
        arr[i].Trigger_key = keys;
    }
    return {arr}
}
```

**Reply by EnzoBitMatrix** [reply]
> Try this:
> 
> ```
> module.exports = async function(data) {
>     var arr = {{$last}}
>     var keys = {{$trigger}}.body.keys;
>     for(let i = 0; i < arr.length; i++) {
>         arr[i].Trigger_key = keys;
>     }
>     return {arr}
> }
> ```

Thank you so much!  Progress -  I am really sorry for bombarding you!
Your code worked perfectly to inject the Trigger_keys into the array.

I can now send that data to the 'Trigger Flow' and receive it, however when I run the 'Create' flow, it adds \ to the Trigger Keys, and therefore, making them more than 36 characters (UUID).

Received payload from your script above:
![image](https://user-images.githubusercontent.com/22970783/225486193-051be932-5b57-4cc2-9157-03182cf0c110.png)

Payload for 'Create':
![image](https://user-images.githubusercontent.com/22970783/225486291-2287e1b4-3e4e-4147-8551-6c4f247f4e16.png)

Result (Options) on Create Log
![image](https://user-images.githubusercontent.com/22970783/225486492-e00a53e2-d46d-403a-ab00-aa9d35ecc500.png)


Any idea's on how I can pass the Trigger key without as a raw value?

I can't express my gratitude enough for your help with this one!

**Reply by EnzoBitMatrix** [reply]
For reference, it looks like it might be the same issue you were experiencing here: https://github.com/directus/directus/issues/13787

**Reply by luciandavidescu** [reply]
try puting in the payload using 'raw code' (click on the `{}` thingy to switch between modes)

**Reply by EnzoBitMatrix** [reply]
Thanks, tried editing the raw formulas, still no luck.
I tried doing a 'Log to Console" with this code, and this is what I get... Still getting slashes in the Trigger_key variable..
![image](https://user-images.githubusercontent.com/22970783/225776273-8f52cf75-6d6b-4fa8-90cb-fdd7a58fd9f7.png)

![image](https://user-images.githubusercontent.com/22970783/225776304-d6a252a2-bb4a-4a55-a650-8bc3e847c9b4.png)
